File: 14.py
#!/usr/bin/env python

from itertools import islice


def digits(n):
    ds = []
    while n > 0:
        n, d = divmod(n, 10)
        ds.insert(0, d)
    return ds if ds else [0]


def advance(recipes, elves):
    return [(i + 1 + recipes[i]) % len(recipes) for i in elves]


def experiment(recipes, elves):
    recipes = recipes[:]
    yield from recipes
    while True:
        current = [recipes[i] for i in elves]
        combined = digits(sum(current))
        recipes.extend(combined)
        yield from combined
        elves = advance(recipes, elves)


def window(size, it):
    win = list(islice(it, size))
    while True:
        yield win
        del win[0]
        win.append(next(it))


if __name__ == '__main__':
    s = open('14.input').read().strip()

    # part 1
    recipes = list(islice(experiment([3, 7], [0, 1]), int(s) + 10))
    print(''.join(str(r) for r in recipes[-10:]))

    # part 2
    pattern = [int(d) for d in s]
    n = 0
    for win in window(len(pattern), experiment([3, 7], [0, 1])):
        if win == pattern:
            print(n)
            break
        n += 1

File: 14.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script --package containers

{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE TypeApplications #-}

import Prelude hiding (drop, length, take)
import Data.Sequence

type Index = Int
type Score = Int
type Elves = [Index]
type Scoreboard = (Seq Score, Elves)

digits :: Int -> Seq Int
digits = go empty where
    go ds 0 = ds
    go ds n = go (n `mod` 10 <| ds) (n `div` 10)

combine :: Seq Score -> Seq Score
combine = digits . sum

ix :: Seq Score -> Int -> Index
ix scores i = i `mod` length scores

at :: Seq Score -> Index -> Score
at scores i = scores `index` ix scores i

addScores :: Seq Score -> Elves -> Seq Score
addScores scores elves = scores >< combine recipes where
    recipes = fromList $ map (at scores) elves

advance :: Seq Score -> Elves -> Elves
advance scores = map (adv scores) where
    adv scores i = ix scores $ i + 1 + scores `index` i

oneRound :: Scoreboard -> Scoreboard
oneRound (scores, elves) = (scores', elves') where
    scores' = addScores scores elves
    elves' = advance scores' elves

-- Hmm, how do I turn this into an infinite series of recipe scores?
-- I want my main to simply do: take 10 $ drop 999 recipes...
roundsUntilLength :: Scoreboard -> Int -> Scoreboard
roundsUntilLength board n
    | length (fst board) >= n = board
    | otherwise = roundsUntilLength (oneRound board) n

nRecipesAfterM :: Scoreboard -> Int -> Int -> Seq Score
nRecipesAfterM scores n m = take n $ drop m $ fst $ roundsUntilLength scores (n + m)

join :: Seq Score -> String
join = concatMap show

main :: IO ()
main = do
    input <- readFile "14.input"
    let scores' = map (\c -> read @Int [c]) $ head $ lines input
    let scores = fromList [3, 7]
    let initial = (scores, [0, 1])
    putStrLn $ join $ nRecipesAfterM initial 10 9
    putStrLn $ join $ nRecipesAfterM initial 10 5
    putStrLn $ join $ nRecipesAfterM initial 10 18
    putStrLn $ join $ nRecipesAfterM initial 10 2018
    -- part 1
    putStrLn $ join $ nRecipesAfterM initial 10 637061
    -- 1513321432 is too low!
    -- part 2
